### ISSUE TITLE
[FEATURE] Socks5 Proxy Server for IP Rotation?

### ISSUE THREAD

**Issue by Nou4r** [kitabisa/mubeng#76]
Is running the proxy server as a SOCKS5 Protocol planned?

Thanks a lot for your work on this project, very useful tool but it would become 10 times more useful if we could run the proxy server as a SOCKS5

**Comment by fuad00** [comment]
any updates/planes on this?

**Comment by dwisiswant0** [comment]
This has already been implemented.

**Comment by llp-czj** [comment]
You can do this if you need to

can use "golang.org/x/net/proxy" lib：

demo : 

`internal/server/server.go`


```golang
import (
  "github.com/things-go/go-socks5"
  netProxy "golang.org/x/net/proxy"
)


func RunSocks5Proxy(opt *common.Options) {
	var recs []*logo.Receiver
	cli := logo.NewReceiver(os.Stderr, "")
	cli.Color = true
	cli.Level = logo.DEBUG
	recs = append(recs, cli)
	file, err := logo.Open(opt.Output)
	if err == nil {
		out := logo.NewReceiver(file, "")
		out.Format = "%s: %s"
		recs = append(recs, out)
	}
	handler = &Proxy{}
	handler.Options = opt
	log = logo.NewLogger(recs...)
	// use netProxy golang.org/x/net/proxy
	socksOpts := []socks5.Option{
		socks5.WithLogger(log),
		// set Dial func
		socks5.WithDial(func(ctx context.Context, net_, addr string) (net.Conn, error) {
			// get one socks proxy uri => socks5://xxxx
			uri := handler.getProxy()
			parse, err := url.Parse(uri)
			if err != nil {
				return nil, err
			}
			if parse.Host == "" {
				return nil, errors.New("socks5 address invalid")
			}
			// processing certification
			password, _ := parse.User.Password()
			auth := &netProxy.Auth{
				User:     parse.User.Username(),
				Password: password,
			}
			// get socks5 dialer
			dialer, err := netProxy.SOCKS5("tcp", parse.Host, auth, netProxy.Direct)
			if err != nil {
				return nil, err
			}
			log.Infof("proxy server: %v target addr: %v", parse.Host, addr)
			return dialer.Dial(net_, addr)
		}),
	}
	if opt.Auth != "" {
		splitN := strings.SplitN(opt.Auth, ":", 2)
		credentials := socks5.StaticCredentials(map[string]string{splitN[0]: splitN[1]})
		socksOpts = append(socksOpts, socks5.WithCredential(credentials))
		log.Infof("SockeAuth Uname:%v Pwd:%v", splitN[0], splitN[1])
	}
	// get socks5 server
	socks5Server := socks5.NewServer(socksOpts...)
	if opt.Watch {
		watcher, err := opt.ProxyManager.Watch()
		if err != nil {
			log.Fatal(err)
		}
		defer watcher.Close()
		go watch(watcher)
	}
	// maintain a healthy proxy pool ( I need it )
	go func() {
		healthyFunc := func() error {
			// 1. check socks proxy alive
			// Use checker.Do() => because checker.check cannot be used directly
			opt.Output = "pipe"
			r, w, err := os.Pipe()
			if err != nil {
				return err
			}
			opt.Result = w
			opt.Verbose = true
			checker.Do(opt)
			w.Close()
			buffer, err := io.ReadAll(r)
			if err != nil {
				return err
			}
			r.Close()
			handler.Options.ProxyManager.Proxies = strings.Split(string(buffer), "\n")
			// 2. clear other type proxy
			for _, proxy := range handler.Options.ProxyManager.Proxies {
				if !strings.Contains(proxy, "socks5") || strings.ReplaceAll(proxy, " ", "") == "" {
					handler.Options.ProxyManager.RemoveProxy(proxy)
				}
			}
			log.Infof("current proxy count: %d", handler.Options.ProxyManager.Count())
			return nil
		}
		for {
			err := healthyFunc()
			if err != nil {
				log.Error(err)
			}
			time.Sleep(30 * time.Minute)
		}
	}()
	log.Infof("%d proxies loaded", opt.ProxyManager.Count())
	log.Infof("[PID: %d] Starting proxy server on %s", os.Getpid(), opt.Address)
	if opt.Auth != "" {
		log.Infof("socks url: socks5://%v:%v", opt.Auth, opt.Address)
	} else {
		log.Infof("socks url: socks5://%v", opt.Address)
	}
	// start socks5 server
	err = socks5Server.ListenAndServe("tcp", opt.Address)
	if err != nil {
		log.Fatal(err)
	}
}

```

getProxy:

```golang
func (p *Proxy) getProxy() string {
	if p.currentProxy == "" || int(p.num) >= p.Options.Rotate {
		proxy, err := p.Options.ProxyManager.Rotate(p.Options.Method)
		if err != nil {
			log.Fatalf("Could not rotate proxy IP: %s", err)
		}
		p.mu.Lock()
		p.currentProxy = proxy
		p.num = 1
		p.mu.Unlock()
	} else {
		atomic.AddUint32(&p.num, 1)
	}
	return p.currentProxy
}
```
runner.New

```golang
func New(opt *common.Options) error {
	if opt.Address != "" {
		if opt.Daemon {
			return daemon.New(opt)
		}
		if opt.Socks {
			log.Println("start mubeng socks5 proxy server mode ...")
			server.RunSocks5Proxy(opt)
		} else {
			log.Println("start mubeng http proxy server mode ...")
			server.RunHTTPProxy(opt)
		}
	} else if opt.Check {
		checker.Do(opt)
		if opt.Output != "" {
			defer opt.Result.Close()
		}
	} else {
		return errors.New("no action to run")
	}
	return nil
}
```

![Image](https://github.com/user-attachments/assets/fa4a60c4-ccc3-48ea-99e5-6eb4dd8f9ef2)

![Image](https://github.com/user-attachments/assets/3ab5c93a-8d2f-46bd-8dd2-ab232bfd37c9)

![Image](https://github.com/user-attachments/assets/2257af75-4f89-4da5-83b2-778afb21b2a8)

**Comment by dwisiswant0** [comment]
Anyone is more than welcome to jump in and work on adding this feature! To help get started, here are some raw specs I’ve put together based on how I think this could be implemented:

1. **first**  
   The type of proxy server could be controlled using a flag, for example, `--type`. This flag would allow users to specify the kind of proxy server they want to run, with the following supported values:  
   - "http" (which would act as the default type if the flag isn’t explicitly provided).  
   - "https" (this would require additional flags related to certificates - blocked by #251).  
   - "socksN" (where $$N$$ represents the version of the SOCKS protocol to be used).  

   The main limitation with this approach is that the proxy server would only be able to serve one type of proxy at a time. So if users need multiple proxy types simultaneously, they would have to run multiple instances of the app, each with its own config.

2. **second option**  
   To address the limitation in the first option, we could introduce a separate `--socks-address` flag. This flag would make it possible to configure and run both an HTTP proxy server and a SOCKS proxy server within the same instance. By specifying an address for the SOCKS server (in addition to the primary address for the HTTP proxy), users could achieve dual functionality w/o needing multiple instances.

In summary, the first approach keeps things simple and focused but lacks flexibility, while the second approach provides a more versatile solution. Either option could work, or both could even be combined to give users greater control. That’s all I've got for now.

**Comment by zi0wi15on** [comment]
Hello, developers, I modified the mubeng source code of v0.22.0 and v0.21.0 by imitating this branch[add support socks5 server type #280](https://github.com/mubeng/mubeng/pull/280)  and I have checked many times to ensure that there are no errors and omissions in the modified code, and then tried to compile it into .exe to execute. The following are my command parameters:

```powershell
.\mubeng.exe -a :7899 --type socks5 -f .\live.txt -r 1 --rotate-on-error  --max-errors -3 --remove-on-error -m random -s -w
``` 

Then I package the test in another terminal:

```powershell
curl --proxy socks5://127.0.0.1:7899 -k https://api.ipify.org
``` 

<img width="3184" height="317" alt="Image" src="https://github.com/user-attachments/assets/41187ff2-2c88-492f-8d05-c5cb5ff7eb13" />

<img width="3187" height="1690" alt="Image" src="https://github.com/user-attachments/assets/75627a1a-b1a1-45d8-a89a-b1b444012dac" />

I found that when the connection error occurred, mubeng did not re-package as expected. I would like to ask which part of the code needs to be adjusted?

**Comment by liancccc** [comment]
@zi0wi15on  Hello, the modified version only supports socks5 but does not add the error reconnection function。You can try to use this way：

<img width="1551" height="436" alt="Image" src="https://github.com/user-attachments/assets/ee61d871-5552-42fb-b88d-224982b1c3c1" />

Since I need to use a proxy over TCP and don't really care about error conditions, I have no plans to modify it.
